contracts/src/TeleporterTokenSource.sol

```diff
@@ -182,9 +317,54 @@ abstract contract TeleporterTokenSource is
         );
 
         if (isMultihop) {
-            emit TokensAndCallRouted(messageID, input, amount);
+            emit TokensAndCallRouted(messageID, input, adjustedAmount);
         } else {
-            emit TokensAndCallSent(messageID, originSenderAddress, input, amount);
+            emit TokensAndCallSent(messageID, originSenderAddress, input, adjustedAmount);
+        }
+    }
+
+    /**
+     * @dev See {INativeTokenSource-addCollateral}
+     */
+    function _addCollateral(
+        bytes32 destinationBlockchainID,
+        address destinationBridgeAddress,
+        uint256 amount
+    ) internal sendNonReentrant {
+        DestinationBridgeSettings memory destinationSettings =
+            registeredDestinations[destinationBlockchainID][destinationBridgeAddress];
+        require(
+            destinationSettings.registered,
+            "TeleporterTokenSource: destination bridge not registered"
+        );
+        require(
+            destinationSettings.collateralNeeded > 0,
+            "TeleporterTokenSource: zero collateral needed"
+        );
+
+        // Deposit the full amount, and withdraw back to the sender if there is excess.
+        amount = _deposit(amount);
+
+        // Calculate the amount remaining.
+        uint256 remainingAmount;
+        uint256 excessAmount;
+        if (amount >= destinationSettings.collateralNeeded) {
+            remainingAmount = 0;
+            excessAmount = amount - destinationSettings.collateralNeeded;
+        } else {
+            remainingAmount = destinationSettings.collateralNeeded - amount;
+        }
+
+        // Update the reserve imbalance remaining.
+        registeredDestinations[destinationBlockchainID][destinationBridgeAddress].collateralNeeded =
+            remainingAmount;
+        emit CollateralAdded(
+            destinationBlockchainID, destinationBridgeAddress, amount, remainingAmount
+        );
+
+        // If there is excess amount, send it back to the sender.
+        if (excessAmount > 0) {
+            _withdraw(msg.sender, excessAmount);
```

Comment:
In the `NativeTokenSource` case, do we need to worry about `msg.sender` not being `payable` here?

Reply:
right now this is treated the same as any other `_withdraw` call that is made to a recipient that cannot accept native tokens, the call would revert, which I think is okay since they can call `addCollateral` again either from an address than can accept tokens, or with an exact collateral amount. As opposed to depositing the extra into bridge